core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSource.java

```diff
@@ -310,33 +208,105 @@ public synchronized void close()
         }
     }
 
-    interface Callback
+    private static class IdempotentSplitSource
+            implements Closeable
     {
-        void partitionsAdded(List<Partition> partitions);
+        private final QueryId queryId;
+        private final TableExecuteContextManager tableExecuteContextManager;
+        private final PlanNodeId planNodeId;
+        private final Optional<PlanFragmentId> sourceFragmentId;
+        private final SplitSource splitSource;
+        private final int splitBatchSize;
+        private final LongConsumer getSplitTimeRecorder;
 
-        void noMorePartitions();
+        @GuardedBy("this")
+        private ListenableFuture<SplitBatchReference> future;
```

Comment:
nit: look like keeping the field as `Optional<>` instead nullable would make code slightly clener. Does not seem from performance perspective.